### DISCUSSION TITLE
Issues rendering annotations from pdf to image

### DISCUSSION THREAD

**Discussion by geirthanas** [pypdfium2-team/pypdfium2#293] [Q&A] [answered]
Hi! I have a pdf with 4 annotations/field forms that are filled out on my pdf, and I wanted to convert it to an image, however when I convert it the image no longer contains the filled out values- I'm using python. 

I've tried specifying it here:
`pil_image = page.render(scale = 300/72, may_draw_forms=True, draw_annots=True).to_pil()
`
and I've also taken from the documentation and tried rendering it through a bitmap but having the same issue:

```python
render_args = (
    bitmap,  # the bitmap
    page,    # the page
    # positions and sizes are to be given in pixels and may exceed the bitmap
    0,       # left start position
    0,       # top start position
    width,   # horizontal size
    height,  # vertical size
    0,       # rotation (as constant, not in degrees!)
    pdfium_c.FPDF_ANNOT,  # rendering flags, combined with binary or
)

# Render the page
pdfium_c.FPDF_RenderPageBitmap(*render_args)
```

Any suggestions on why it might not be rendering in the image?

**Comment by mara004** [comment, marked as answer]
On the support model example: Make sure to call `.init_forms()` on the parent pdf before the page is instantiated.

On the raw API example, you'd need to call `FPDF_FFLDraw(...)`, which takes an `FPDF_FORMHANDLE` and the render args.
Setting up the form requires several other APIs, which is somewhat complicated.
The support model should be more straightforward for your use case.

I agree the Readme's raw API rendering example is incomplete when it comes to forms.
The support model's implementation is more exhaustive:

*init_forms()*
https://github.com/pypdfium2-team/pypdfium2/blob/a5849277c087c932bfb9187242a80cd279e1bdab/src/pypdfium2/_helpers/document.py#L150-L183
*get_page()*
https://github.com/pypdfium2-team/pypdfium2/blob/a5849277c087c932bfb9187242a80cd279e1bdab/src/pypdfium2/_helpers/document.py#L367-L379
*render()*
https://github.com/pypdfium2-team/pypdfium2/blob/a5849277c087c932bfb9187242a80cd279e1bdab/src/pypdfium2/_helpers/page.py#L439-L440

**Reply by mara004** [reply]
In addition, if your document uses an XFA form, you'd need a special build of pdfium.
In that case, `init_forms()` would give you a warning with instructions how to procure such a build.

**Reply by geirthanas** [reply]
Thank you this worked for me, getting the FPDF_FFLDraw added to my calls made it work! Thank you!
